systemd/provider.go

```diff
@@ -83,18 +87,29 @@ func New(cfg provider.InitConfig) (*P, error) {
 
 	p.Host = restConfig.Host
 
+	// Set-up a clientset to be used in Secret and ConfigMap reconciliation.
 	clientset, err := nodeutil.ClientsetFromEnv(cfg.ConfigPath)
 	if err != nil {
 		return p, err
 	}
-	// Get new clientset.
-	w := newWatcher(clientset)
-	go func() {
-		if err := w.run(p); err != nil {
-			klog.Fatal(err)
-		}
-	}()
-	p.w = w
+
+	// Set the event handler functions for Secrets and ConfigMaps.
+	p.w = newWatcher()
+
+	// Set-up informers and extract listers to use when accessing Kubernetes resources.
+	informerFactory := informers.NewSharedInformerFactory(clientset, time.Minute*1)
+
+	secretInformer := informerFactory.Core().V1().Secrets()
+	secretInformer.Informer().AddEventHandler(p.w.handlerFuncs(p))
+	p.secretLister = secretInformer.Lister()
+
+	cmInformer := informerFactory.Core().V1().ConfigMaps()
+	cmInformer.Informer().AddEventHandler(p.w.handlerFuncs(p))
+	p.cmLister = cmInformer.Lister()
+
+	informerFactory.Start(wait.NeverStop)
+	informerFactory.WaitForCacheSync(wait.NeverStop)
```

Comment:
should we just move this into `newWatcher`?

Reply:
I tried to keep the Watcher as clean as possible and in so it expects solely a cache which in the future may or may not depend on the setup of an informer.